### ISSUE TITLE
NPE in TopicFragmentFilter [DITA OT 2.0]

### ISSUE THREAD

**Issue by raducoravu** [dita-ot/dita-ot#1902]
Based on [this longer discussion](https://groups.yahoo.com/neo/groups/dita-users/conversations/messages/37299). If I have a DITA topic on which the body has been marked by mistake as another topic:

``` xml
<!DOCTYPE topic PUBLIC "-//OASIS//DTD DITA Topic//EN" "http://docs.oasis-open.org/dita/v1.1/OS/dtd/topic.dtd">
<topic id="introduction">
   <title>Introduction</title>
  <body class="- topic/topic ">
```

Transformation fails with:

> BUILD FAILED
> C:\Users\radu_coravu\Desktop\DOT2.0.1\DITA-OT2.0.1\build.xml:41: The following error occurred while executing this line:
> C:\Users\radu_coravu\Desktop\DOT2.0.1\DITA-OT2.0.1\plugins\org.dita.base\build_preprocess.xml:244: java.lang.NullPointerException
>     at java.util.ArrayDeque.addFirst(Unknown Source)
>     at org.dita.dost.writer.TopicFragmentFilter.startElement(TopicFragmentFilter.java:36)
>     at org.apache.xerces.parsers.AbstractSAXParser.startElement(Unknown Source)
>     at org.ditang.relaxng.defaults.RelaxNGDefaultsComponent.startElement(RelaxNGDefaultsComponent.java:200)
>     at org.apache.xerces.impl.XMLNSDocumentScannerImpl.scanStartElement(XMLNSDocumentScannerImpl.java:268)
>     at org.apache.xerces.impl.XMLDocumentFragmentScannerImpl$FragmentContentDispatcher.dispatch(XMLDocumentFragmentScannerImpl.java:1655)
>     at org.apache.xerces.impl.XMLDocumentFragmentScannerImpl.scanDocument(XMLDocumentFragmentScannerImpl.java:325)
>     at org.apache.xerces.parsers.XML11Configuration.parse(Unknown Source)
>     at org.ditang.relaxng.defaults.RelaxDefaultsParserConfiguration.parse(RelaxDefaultsParserConfiguration.java:150)
>     at org.apache.xerces.parsers.XML11Configuration.parse(Unknown Source)
>     at org.apache.xerces.parsers.XMLParser.parse(Unknown Source)
>     at org.apache.xerces.parsers.AbstractSAXParser.parse(Unknown Source)
>     at org.xml.sax.helpers.XMLFilterImpl.parse(Unknown Source)
>     at net.sf.saxon.event.Sender.sendSAXSource(Sender.java:441)
>     at net.sf.saxon.event.Sender.send(Sender.java:146)
>     at net.sf.saxon.jaxp.IdentityTransformer.transform(IdentityTransformer.java:363)
>     at org.dita.dost.util.XMLUtils.transform(XMLUtils.java:188)
>     at org.dita.dost.util.XMLUtils.transform(XMLUtils.java:148)
>     at org.dita.dost.module.TopicFragmentModule.execute(TopicFragmentModule.java:49)

**Comment by BenAllums** [comment]
Experiencing the same issue on DITA-OT 2.1.1.

```
BUILD FAILED
Q:\@factory\ePublisher_Platform\trunk\dev\source\windows\dotnet\WebWorks.test\Publish\Core\ProgramDirectory\Helpers\DITA-O~1.1\plugins\org.dita.base\build_preprocess.xml:248: java.lang.NullPointerException
	at java.util.ArrayDeque.addFirst(Unknown Source)
	at org.dita.dost.writer.TopicFragmentFilter.startElement(TopicFragmentFilter.java:36)
	at org.apache.xerces.parsers.AbstractSAXParser.startElement(Unknown Source)
	at org.apache.xerces.impl.XMLNSDocumentScannerImpl.scanStartElement(Unknown Source)
	at org.apache.xerces.impl.XMLNSDocumentScannerImpl$NSContentDispatcher.scanRootElementHook(Unknown Source)
	at org.apache.xerces.impl.XMLDocumentFragmentScannerImpl$FragmentContentDispatcher.dispatch(Unknown Source)
	at org.apache.xerces.impl.XMLDocumentFragmentScannerImpl.scanDocument(Unknown Source)
	at org.apache.xerces.parsers.XML11Configuration.parse(Unknown Source)
	at org.apache.xerces.parsers.XML11Configuration.parse(Unknown Source)
	at org.apache.xerces.parsers.XMLParser.parse(Unknown Source)
	at org.apache.xerces.parsers.AbstractSAXParser.parse(Unknown Source)
	at org.xml.sax.helpers.XMLFilterImpl.parse(Unknown Source)
	at net.sf.saxon.event.Sender.sendSAXSource(Sender.java:404)
	at net.sf.saxon.event.Sender.send(Sender.java:193)
	at net.sf.saxon.IdentityTransformer.transform(IdentityTransformer.java:30)
	at org.dita.dost.util.XMLUtils.transform(XMLUtils.java:188)
	at org.dita.dost.util.XMLUtils.transform(XMLUtils.java:148)
	at org.dita.dost.module.TopicFragmentModule.execute(TopicFragmentModule.java:50)
	at org.dita.dost.pipeline.PipelineFacade.execute(PipelineFacade.java:68)
	at org.dita.dost.invoker.ExtensibleAntInvoker.execute(ExtensibleAntInvoker.java:193)
	at org.apache.tools.ant.UnknownElement.execute(UnknownElement.java:293)
	at sun.reflect.GeneratedMethodAccessor4.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
	at java.lang.reflect.Method.invoke(Unknown Source)
	at org.apache.tools.ant.dispatch.DispatchUtils.execute(DispatchUtils.java:106)
	at org.apache.tools.ant.Task.perform(Task.java:348)
	at org.apache.tools.ant.Target.execute(Target.java:435)
	at org.apache.tools.ant.Target.performTasks(Target.java:456)
	at org.apache.tools.ant.Project.executeSortedTargets(Project.java:1405)
	at org.apache.tools.ant.Project.executeTarget(Project.java:1376)
	at org.apache.tools.ant.helper.DefaultExecutor.executeTargets(DefaultExecutor.java:41)
	at org.apache.tools.ant.Project.executeTargets(Project.java:1260)
	at org.apache.tools.ant.Main.runBuild(Main.java:853)
	at org.apache.tools.ant.Main.startAnt(Main.java:235)
	at org.apache.tools.ant.launch.Launcher.run(Launcher.java:285)
	at org.apache.tools.ant.launch.Launcher.main(Launcher.java:112)
```

**Comment by robander** [comment]
I'm very nervous about this (two aspects of it).

First - the `@id` attribute on `<topic>` is an absolute requirement of the DITA grammar, at the root of pretty much all addressing rules / processing. Removing the check for `@id` will allow the topic to pass that step, and might give reasonable output when processed in isolation, but when mixed with other content it will give strange errors / results from many other programs in the tool chain that expect `@id`. For the simple case of a real `<topic>` that is actually missing `@id`, I think it would be better to generate one in order to make all of the later processing steps more robust.

For the original problem, with `<body class="- topic/topic ">`, we have a `<body>` element that is telling any DITA processor it is actually a `<topic>`. Here too I would expect failures from lots of other steps, depending on what else is in the topic / what else is in the map. Many of our processing steps are based on working within a topic, nested topic, etc. For example, map linking, conref, `@ID` generation in output, metadata pushing and pulling -- all of these will look for ` topic/topic `, and will probably give bad output when the `<body>` claims to be a topic. So I'm not sure how kind we are being by letting that through the first step? (Honest question there, I'm really not sure.)

**Comment by jelovirt** [comment]
Fixed by 23f980c
